tt_metal/impl/kernels/runtime_args_data.cpp

```diff
@@ -0,0 +1,44 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "tt_metal/impl/kernels/runtime_args_data.hpp" // Public API Header
+
+#include "tt_metal/common/assert.hpp"
+
+namespace tt::tt_metal {
+
+std::uint32_t & RuntimeArgsData::operator[](std::size_t index) {
+    TT_ASSERT(index < rt_args_count, "Index specified is larger than runtime args size");
+    return this->rt_args_data[index];
+}
+
+const std::uint32_t & RuntimeArgsData::operator[](std::size_t index) const {
+    TT_ASSERT(index < rt_args_count, "Index specified is larger than runtime args size");
+    return this->rt_args_data[index];
+}
```

Comment:
Another potential follow-up is to remove the bounds check for `operator[]` and make it `noexcept`, having a bounds check here is surprising, and makes it unnecessary to have both `operator[]` and `at` implemented the same way (spoiler alert, `TT_ASSERT` and `TT_FATAL` are effectively equivalent, `std::unreachable();` statement after a `throw` does not make it behave or optimize any differently)

Reply:
TT_ASSERT only runs in debug builds, whereas TT_FATAL will still trigger in release builds, but I agree we can remove the TT_ASSERT from `operator[]`